### PR TITLE
Fix a crash when there are submarkets, but no storageData to render

### LINKED ISSUE #58: Encountered a crash with array-out-of-bounds when trying to open the market selection UI
Linux
Pop!_OS 20.04

Also posted on the forum

```
ERROR com.fs.starfarer.combat.CombatMain  - java.lang.ArrayIndexOutOfBoundsException: -1
java.lang.ArrayIndexOutOfBoundsException: -1
	at java.util.ArrayList.elementData(Unknown Source)
	at java.util.ArrayList.remove(Unknown Source)
	at stelnet.storage.view.StorageTabViewFactory.addStorageData(StorageTabViewFactory.java:101)
	at stelnet.storage.view.StorageTabViewFactory.getTabPane(StorageTabViewFactory.java:73)
	at stelnet.storage.view.StorageTabViewFactory.createContainer(StorageTabViewFactory.java:45)
	at stelnet.market.MarketViewBoard.getRenderables(MarketViewBoard.java:64)
	at stelnet.ui.RenderableIntel.createLargeDescription(RenderableIntel.java:51)
	at com.fs.starfarer.campaign.comms.v2.EventsPanel.Ò00000(Unknown Source)
	at com.fs.starfarer.campaign.comms.v2.EventsPanel.actionPerformed(Unknown Source)
	at com.fs.starfarer.ui.newnew.buttonPressed(Unknown Source)
	at com.fs.starfarer.ui.oooOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO.Ò00000(Unknown Source)
	at com.fs.starfarer.ui.oooOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO.processInput(Unknown Source)
	at com.fs.starfarer.ui.newnew.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.Oo0o$Oo.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Oo0o.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.campaign.comms.v2.EventsPanel.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.campaign.comms.OOoO.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.newui.class.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Oo0O.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.newui.o0oO.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Oo0O.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.newui.Stringsuper.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.ui.oO00.dispatchEventsToChildren(Unknown Source)
	at com.fs.starfarer.ui.oO00.processInputImpl(Unknown Source)
	at com.fs.starfarer.ui.Q.processInput(Unknown Source)
	at com.fs.starfarer.campaign.CampaignState.processInput(Unknown Source)
	at com.fs.starfarer.BaseGameState.traverse(Unknown Source)
	at com.fs.state.AppDriver.begin(Unknown Source)
	at com.fs.starfarer.combat.CombatMain.main(Unknown Source)
	at com.fs.starfarer.StarfarerLauncher.o00000(Unknown Source)
	at com.fs.starfarer.StarfarerLauncher$1.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
```

### PR DESCRIPTION
**Pull request by josiah-roberts**
Fixes #58.

I tested this on my crashing save, and it no longer crashes. Poked around opening various boards, looking at different markets, etc -- everything seems to work as expected.

I don't have much context on this codebase, so there might be a more elegant/clean way to handle this edge case.

### CODE REVIEW

**Comment by jaghaimo** [comment]
Thanks for the report and fix. I'll have a look at it on Monday.

Are you the same person that reported it on the forums?

File: src/main/java/stelnet/storage/view/StorageTabViewFactory.java [hunk 1]
```diff
@@ -89,7 +89,7 @@ private void addEmptyData(List<AbstractRenderable> elements, List<StorageData> s
     }
 
     private void addStorageData(List<AbstractRenderable> elements, List<StorageData> storageData) {
-        if (!hasStorage()) {
+        if (!hasStorage() || storageData.isEmpty()) {
```
Comment (jaghaimo) [inline src/main/java/stelnet/storage/view/StorageTabViewFactory.java:92]: What confuses me is this check should not be needed for iterating over Collection. If it is empty, then for each will not run.

Comment (jaghaimo) [reply src/main/java/stelnet/storage/view/StorageTabViewFactory.java:92]: Found the bug. It's actually the code to remove last spacer element, which does not exist if it is empty.

**Comment by jaghaimo** [comment]
I will be going with removal of protected method `hasStorage`, change of `addStorageData` to:
```java
    private void addStorageData(List<AbstractRenderable> elements, List<StorageData> storageData) {
        for (StorageData data : storageData) {
            LocationData locationData = data.getLocationData();
            elements.add(new Heading(locationData.getName(), locationData.getFgColor(), locationData.getBgColor()));
            elements.add(activeTab.getStorageRenderer(data));
            elements.add(new Spacer(8));
        }
        removeLastElement(elements);
    }

    private void removeLastElement(List<AbstractRenderable> elements) {
        if (elements.size() > 0) {
            elements.remove(elements.size() - 1);
        }
    }
```
And changing check in `addEmptyData` to just `storageData.isEmpty()`. Could you check this also fixes your issue, before I close it?

https://github.com/jaghaimo/stelnet/pull/60

**Comment by josiah-roberts** [comment]
Hey @jaghaimo - sorry my Github notifications are kind of a perpetual thunderstorm. I'll take a look tonight and check if my issue is fixed on latest.

**Pull request merged**